src/compile.rs

```diff
@@ -119,23 +119,51 @@ pub fn extend_type_map(p: &PolicyFile, type_map: &mut TypeMap) -> Result<(), Cas
     // TODO: This only allows declarations at the top level.
     // Nested declarations are legal, but auto-associate with the parent, so they'll need special
     // handling when association is implemented
+
     let mut errors = CascadeErrors::new();
     for e in &p.policy.exprs {
         let d = match e {
             Expression::Decl(d) => d,
             _ => continue,
         };
-        match d {
-            Declaration::Type(t) => {
-                if !t.is_extension {
-                    match TypeInfo::new(*t.clone(), &p.file) {
-                        Ok(new_type) => type_map.insert(t.name.to_string(), new_type)?,
-                        Err(e) => errors.append(e),
+        if let Declaration::Type(t) = d {
+            // If there are nested declarations, they associate
+            let mut associate_names = BTreeSet::new();
+            for e in &t.expressions {
+                if let Expression::Decl(Declaration::Type(associated_type)) = e {
+                    // Make the synthetic type to associate
+                    if type_map.get(associated_type.name.as_ref()).is_none() {
+                        match TypeInfo::new(*associated_type.clone(), &p.file) {
+                            Ok(new_type) => {
+                                type_map.insert(associated_type.name.to_string(), new_type)?
+                            }
+                            Err(e) => errors.append(e),
+                        }
                     }
+                    associate_names.insert(associated_type.name.clone());
                 }
             }
-            _ => continue,
-        };
+
+            if !t.is_extension {
+                match TypeInfo::new(*t.clone(), &p.file) {
+                    Ok(mut new_type) => {
+                        if !associate_names.is_empty() {
+                            // TODO: this returns false if it already existed.  That should
+                            // probably be an error
+                            new_type
+                                .annotations
+                                .insert(AnnotationInfo::Associate(Associated {
+                                    resources: associate_names,
+                                }));
+                        }
+                        type_map.insert(t.name.to_string(), new_type)?;
+                    }
+                    Err(e) => errors.append(e),
+                }
+            } else if !associate_names.is_empty() {
+                todo!()
```

Comment:
May want to add a comment here just so we don't forget what we wanted todo here.